# Single Nucleotide Polymorphism in Chinese Sassafras Populations

## Background
Since SNPs can help identify genes that are important and that differ between organisms, we can use them to identify important gene differences that may be due to environmental conditions. The research question is: Can we use machine learning algorithms to mine SNPs to find gene or gene regions of interest between natural cultivars (strains) of Sassafras? This question addresses the need for protecting vulnerable populations of Sassafras that are highly sought after due to their ornamental, lumber, and medicinal value (Guan, et al., 2024). It has also been suggested that climate change will have an impact on the habitat availability for Sassafras in China (Zhang, et al., 2020). Future research could lead to a drought or cold resistant strain of Sassafras for industrial uses. 
Although numerous methods to study SNPs and identify gene regions exist, none of these methods are without their drawbacks. SNPs are notoriously difficult to study and machine learning techniques may provide new insight towards this problem. One issue is that genetic data is usually very high dimensional data, leading to the ‘curses of dimensionality’; high dimensional data is often computationally costly and may not yield the best results as not all of the data is useful to our aims (Silva et al., 2022). Using methods such as k-fold cross-validation can help create more accurate models and prevent overfitting. Using variables from the environment these plants were collected, including altitude gleaned from their latitude and longitude, weather data such as temperature, precipitation, humidity, and average days of sunshine, and possible data about soil that we can find from the latitudes and longitudes, we will attempt to find SNPs, and therefore gene or gene regions, that predict key differences between Sassafras cultivars.  These results could lead to actionable insights that could guide future research towards protecting this species and may even be able to improve forestry approaches towards minimizing the impacts of climate change and deforestation.


## Hypothesis and Prediction

Hypothesis: Underlying genes, as identified by SNPs, in Sassafras are influenced by environmental factors because environmental pressure can cause mutations to persist in a population that is unique to each area.

Prediction: Populations of Sassafras that are under high environmental pressure are more likely to have many predictive SNPs due to evolutionary influences. 


## Basic Data

```{r}
install.packages("vcfR", verbose = FALSE)
```
```{r}
library(vcfR)
```
```{r}
vcf <- read.vcfR( "Data/SNP.vcf", verbose = FALSE )
vcf
```

```{r}
queryMETA(vcf)
queryMETA(vcf, element = 'GT')
queryMETA(vcf, element = 'DP')
queryMETA(vcf, element = 'AD')
queryMETA(vcf, element = 'GL')
```

```{r}

vcf_field_names(vcf)
```
```{r}
#These are the samples
colnames(vcf@gt)
```


```{r}
data <- vcfR2tidy(vcf)
```
```{r}
data$meta
```


```{r}
#Summary of the all the SNPs
data$fix
```

```{r}
#Summary of the all the samples
data$gt
```
```{r}
install.packages("adegenet")
library(adegenet)
```


```{r}
x <- vcfR2genlight(vcf)
x
```

```{r}
x.dist <- dist(x)
glPlot(x)
```

```{r}
myFreq <- glMean(x)
hist(myFreq, proba=TRUE, col="gold", xlab="Allele frequencies",
main="Distribution of (second) allele frequencies")
temp <- density(myFreq)
lines(temp$x, temp$y*1.5,lwd=2)
```


```{r}
myFreq <- glMean(x)
myFreq <- c(myFreq, 1-myFreq)
hist(myFreq, proba=TRUE, col="darkseagreen3", xlab="Allele frequencies",
main="Distribution of allele frequencies", nclass=20)
temp <- density(myFreq, bw=.05)
lines(temp$x, temp$y*2,lwd=3)
```

```{r}
temp <- density(glNA(x), bw=10)
plot(temp, type="n", xlab="Position in the alignment", main="Location of the missing val", xlim=c(0,1701))
polygon(c(temp$x,rev(temp$x)), c(temp$y, rep(0,length(temp$x))), col=transp("blue",.3))
points(glNA(x), rep(0, nLoc(x)), pch="|", col="blue")
```
```{r}
# An empty plot.
freq_peak_plot(pos=1:40)
gt <- extract.gt(vcf)
hets <- is_het(gt)
# Censor non-heterozygous positions.
is.na(vcf@gt[,-1][!hets]) <- TRUE
# Extract allele depths.
ad <- extract.gt(vcf, element = "AD")
ad1 <- masplit(ad, record = 1)
ad2 <- masplit(ad, record = 2)
freq1 <- ad1/(ad1+ad2)
freq2 <- ad2/(ad1+ad2)
myPeaks1 <- freq_peak(freq1, getPOS(vcf))
is.na(myPeaks1$peaks[myPeaks1$counts < 20]) <- TRUE
myPeaks2 <- freq_peak(freq2, getPOS(vcf), lhs = FALSE)
is.na(myPeaks2$peaks[myPeaks2$counts < 20]) <- TRUE
freq_peak_plot(pos = getPOS(vcf), ab1 = freq1, ab2 = freq2, fp1 = myPeaks1, fp2=myPeaks2)
```


```{r}
#ChromR object is not working
#chrom <- create.chromR(name="Supercontig", vcf=vcf, verbose=FALSE)
```

```{r}
chrom
```
```{r}
plot(chrom)
```

